### PR TITLE
Add Expo debugging without mandatory logging in and Internet connection

### LINKED ISSUE #1188: Allow expo debugging without logging in / without an internet connection
Feature request to allow debugging expo without expo login or internet connection. 

I understand that expo uses an account and an internet connection for some features, but debugging does not use any of expo's cloud features. 

The expo cli allows development without logging in and without an active internet connection. 

## Actual Behavior
1. Run Debug in Exponent
2. Expo prompts login
3. a. Esc or leave username and password blank
OR
b. Internet connection does not exist
4. Debugging fails to launch

## Expected Behavior
1. Run Debug in Exponent
2. Expo prompts login
3. a. Esc or leave username and password blank
OR
b. Internet connection does not exist
4. Debugging launches on LAN (or localhost)

## Software versions
- React-Native VS Code extension version: 0.13.2
- VSCode version: 1.41.1
- OS platform and version: macOS 10.15.2
- NodeJS version: 13.3.0
- React Native version: 2.0.1
- Expo SDK version (include if relevant): 36.0.0, expo-cli@3.11.2

## Outputs (Include if relevant)
- Output of the Debug Console (View -> Toggle Debug Console):
3. a.
```
[Error] ApiV2Error: Must provide a username and password to log in.
```
3. b.
```
[Error] Error: getaddrinfo ENOTFOUND exp.host
```

### PR DESCRIPTION
**Pull request by RedMickey**
Fixes #1188

### CODE REVIEW

File: src/extension/exponent/exponentPlatform.ts [hunk 1]
```diff
@@ -18,16 +18,25 @@ const localize = nls.loadMessageBundle();
...
 
-    constructor(runOptions: IRunOptions, platformDeps: MobilePlatformDeps = {}) {
+    constructor(runOptions: IExponentRunOptions, platformDeps: MobilePlatformDeps = {}) {
         super(runOptions, platformDeps);
         this.exponentHelper = new ExponentHelper(runOptions.workspaceRoot, runOptions.projectRoot);
         this.exponentTunnelPath = null;
     }
 
+    public static stopExpoAdbReverse() {
+        if (!ExponentPlatform.isExpoAdbReverseEnable) return;
```
Comment (SounD120) [inline src/extension/exponent/exponentPlatform.ts]:
```suggestion
        if (!ExponentPlatform.isExpoAdbReverseEnable) {
            return;
        }
```

Comment (RedMickey) [reply src/extension/exponent/exponentPlatform.ts]: This part of the code has been removed.

File: src/extension/exponent/exponentPlatform.ts [hunk 2]
```diff
@@ -97,6 +100,33 @@ export class ExponentPlatform extends GeneralMobilePlatform {
         });
     }
 
+    public loginToExponentOrSkip(expoHostType?: "tunnel" | "lan" | "local") {
+        if (expoHostType !== "tunnel") return  Q({});
```
[the remark this answers is not in the document]
Comment (RedMickey) [reply src/extension/exponent/exponentPlatform.ts]: Fixed this code style issue.

File: src/extension/exponent/exponentPlatform.ts [hunk 3]
```diff
@@ -39,41 +39,44 @@ export class ExponentPlatform extends GeneralMobilePlatform {
...
-                            .then(answer => {
-                                if (answer === cancelButton) {
-                                    reject(ErrorHelper.getInternalError(InternalErrorCode.UserCancelledExpoLogin));
-                                }
-                                resolve("");
-                            }, reject);
-                    });
-                }
-            )
+            return this.loginToExponentOrSkip(this.runOptions.expoHostType)
```
[the remark this answers is not in the document]
Comment (RedMickey) [reply src/extension/exponent/exponentPlatform.ts]: We can check that the following way as shown on the attached screenshot. Anyway we should start sequence of promises and I suppose that the previous variant of using of the separate function looks cleaner. Please correct me if I'm wrong.
image

**Pull request merged**